tree/ntuple/v7/doc/architecture.md

```diff
@@ -3,7 +3,439 @@ RNTuple Code Architecture
+For instance, there is no API in the page sink to write an entry, but only to write pages of columns.
+The higher-level APIs, e.g. `RField`, `REntry`, `RNTupleWriter`, take care of presenting the available data as entries where necessary.
+
+The page source also gives access to an `RNTupleDescriptor` through a read/write lock guard.
+The `RNTupleDescriptor` owned by the page source changes only when new cluster meta-data are loaded.
+The header and the cluster group summary information is stable throughout its lifetime (cf. format specification).
+
+### R{NTuple,Field,Column,Cluster,...}Descriptor
+The descriptor classes provide read-only access to the on-disk meta-data of an RNTuple.
+The meta-data include the schema (fields and columns), information about clusters and the page locations.
+The descriptor classes are closely related to the format specification.
+
+For normal read and write tasks, access to the descriptor is not necessary.
+One notable exception is bulk reading, where the descriptor can be used to determine entry boundaries of clusters.
+The descriptors are used internally, e.g. to build an RNTupleModel from the on-disk information.
+The descriptors are also useful for inspection purposes.
+
+The descriptor classes contain a copy of the meta-data; they are not linked to an open page source.
+A descriptor can be used after its originating page source has been deleted.
+
+### RField<T>
+The RField<T> classes are central in RNTuple:
+they link the in-memory representation of data types to their on-disk representation.
+All field classes inherit from `RFieldBase`.
+
+Every type with RNTuple I/O supported has a corresponding RField<T> template specialization.
+Complex types are composed of fields and sub fields.
+E.g., a `struct` is represented by a parent field for the `struct` itself and a subfield for every member of the `struct`.
+Fields of complex types have type-erased versions in addition to the templated ones (e.g., `RVectorField`, `RClassField`).
+In this way, fields can be constructed even if the type information is only available at runtime.
+To this end, `RFieldBase::Create()` creates an `RField` object from a type string.
+
+On the "in-memory" side, fields can construct and destroy objects of their wrapped type
+(cf. `CreateValue()`, `CreateObject()`, `GetDeleter()` methods).
+Existing objects in memory can be bound to fields (cf. `BindValue()` method).
+
+On the "on-disk" side, fields know about the possible column representations of their wrapped type.
+Upon connecting a field to a page source or page sink,
+fields create `RColumn` objects and register them with the page source/sink.
+When reading and writing data, the field maps the in-memory information of an object to read/write calls on its columns.
+For instance, when writing a `std::vector<float>`,
+the field writes to an index column (storing information about the size of the vector), and to a float column (the actual values).
```

Comment:
not sure if important here: the `vector` field has a subfield for the `float`s which in turn handles the column.